app/lib/common/utils/language.dart

```diff
@@ -0,0 +1,32 @@
+import 'package:acter/features/settings/model/language_model.dart';
+import 'package:acter/features/settings/providers/settings_providers.dart';
+import 'package:acter_flutter_sdk/acter_flutter_sdk.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+const languagePrefKey = 'a3.language';
+
+Future<void> initLanguage(WidgetRef ref) async {
+  final prefLanguageCode = await getLanguage();
+  final deviceLanguageCode = PlatformDispatcher.instance.locale.languageCode;
+  final bool isLanguageContain = LanguageModel.allLanguagesList
+      .contains(LanguageModel.fromCode(deviceLanguageCode));
+  if (prefLanguageCode == null && isLanguageContain) {
+    await setLanguage(deviceLanguageCode);
+    ref.read(languageProvider.notifier).update((state) => deviceLanguageCode);
+  } else if (prefLanguageCode != null) {
+    ref.read(languageProvider.notifier).update(
+          (state) => prefLanguageCode,
+        );
+  }
```

Comment:
yeah, I don't think that is the logic we should do. If the user changed the system language between restarts and we have support for that language, we should use that, unless the user has explicitly set the language.

IMHO the logic should be:
1. user has set a language manually, use that
2. pick system language if available
3. fallback to english

But do not implicitly _write_ a language selection upon first start. Why would that be useful?

Reply:
It is the same logic.


> 1) user has set a language manually, use that:

This use-case is satisfied with below code. If user has set language manually then `prefLanguageCode` will not be `null` and below case will executed.

```
else if (prefLanguageCode != null) {
    ref.read(languageProvider.notifier).update(
          (state) => prefLanguageCode,
        );
}
```
  

> 2) Pick system language if available

This use-case is satisfied with below code. If user has not set language manually then `prefLanguageCode` will be `null` and if system language is available then below case will executed.

```
if (prefLanguageCode == null && isLanguageContain) {
    await setLanguage(deviceLanguageCode);
    ref.read(languageProvider.notifier).update((state) => deviceLanguageCode);
} 
```

> 3) fallback to english

If above two use-cases are not satisfied then we are using default "English" language with below code where we have already set 'en' as default language.

`final languageProvider = StateProvider<String>((ref) => 'en');
`

Reply:
This is only sufficient on first-start. If you changed the language _after_ it will not fallback to the system language. I don't understand why we'd want to force _write_ a local config if the user never changed it, it also logically doesn't make any sense to me to do that. What's the benefit?

Reply:
I have followed the WhatsApp mechanism. Let me explain with the example.

Current Status:
- - -
Mobile OS : English Lang
WhatsApp : English Lang

STEP-1 : I have opened the `WhatsApp` and changed the lang to "Hindi". 
So now,
Mobile OS : English Lang
WhatsApp : Hindi Lang

STEP-1 : I have opened the `Mobile App Lang` settings and changed the lang to "Gujarati". 
So now,
Mobile OS : Gujarati Lang
WhatsApp : Hindi Lang (This is preserved here based on the my prior selection)

Reply:
That's not what I have a problem with. [Here is a PR explaining what I want to have changed](https://github.com/acterglobal/a3/pull/1521).